kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java

```diff
@@ -415,36 +462,46 @@ private synchronized void manageConfiguredInterface(Device device, String device
         logger.info("Settings iface \"{}\":{}", deviceId, deviceType);
 
         if (interfaceStatus == KuraInterfaceStatus.DISABLED) {
-            disable(device);
+            disable(device, deviceId);
         } else if (interfaceStatus == KuraInterfaceStatus.UNMANAGED) {
             logger.info("Iface \"{}\" set as UNMANAGED in Kura. Skipping configuration.", deviceId);
         } else { // NMDeviceEnable.ENABLED
             enableInterface(deviceId, properties, device, deviceType);
         }
 
         // Manage GPS independently of device ip status
-        if (deviceType == NMDeviceType.NM_DEVICE_TYPE_MODEM) {
+        if (deviceType == NMDeviceType.NM_DEVICE_TYPE_MODEM && device.isPresent()) {
             Optional<Boolean> enableGPS = properties.getOpt(Boolean.class, "net.interface.%s.config.gpsEnabled",
                     deviceId);
-            Optional<String> mmDbusPath = this.networkManager.getModemManagerDbusPath(device.getObjectPath());
+            Optional<String> mmDbusPath = this.networkManager.getModemManagerDbusPath(device.get().getObjectPath());
             this.modemManager.setGPS(mmDbusPath, enableGPS);
         }
 
     }
 
-    private void enableInterface(String deviceId, NetworkProperties properties, Device device, NMDeviceType deviceType)
+    private void enableInterface(String deviceId, NetworkProperties properties, Optional<Device> optDevice, NMDeviceType deviceType)
```

Comment:
I think this is way more complicated than it needs to be.

The logic is the following:
- If there's no device we just need to create the connection (since when the connection gets created the virtual device will spawn too)
- If there's no device there cannot be an associated connection for which we need to update the settings
- If there's no device there cannot be any additional associated connections to cleanup in the housekeeping stage

Therefore this should be simplified as follows:

```java
    private void enableInterface(String deviceId, NetworkProperties properties, Optional<Device> device, NMDeviceType deviceType)
            throws DBusException {
        if(device.isPresent()) {
            enableInterface(deviceId, properties, device.get(), deviceType);
        } else {
            enableVirtualInterface(deviceId, properties, deviceType);
        }
    }

    private void enableVirtualInterface(String deviceId, NetworkProperties properties, NMDeviceType deviceType)
            throws DBusException {
        Map<String, Map<String, Variant<?>>> newConnectionSettings = NMSettingsConverter.buildSettings(properties,
                Optional.empty(), deviceId, deviceId, deviceType);
        try {
            this.networkManager.AddAndActivateConnection(
                    newConnectionSettings, new DBusPath("/"), new DBusPath("/"));
        } catch (DBusExecutionException e) {
            logger.warn("Couldn't complete activation of {} interface, caused by:", deviceId, e);
        }
    }

    private void enableInterface(String deviceId, NetworkProperties properties, Device device, NMDeviceType deviceType)
            throws DBusException {
        // this is the old method without any change
```

> **Warning**
Here I'm assuming that, since we're creating and activating a connection on a device that not yet exists, we need to pass to the `AddAndActivateConnection` method a null dbus. I didn't have the time to test this assumption though.

Additionally here there's no signal that we can wait for to understand if the connection was applied (since there's no device yet). A possible solution would be to wait for the `DeviceAdded` signal?

Reply:
Mh... apparently this fails with:

```
2023-09-05T08:41:01,589 [ConfigurationListener Event Queue] WARN  o.e.k.n.NMDbusConnector - Couldn't complete activation of eth0.40 interface, caused by:
org.freedesktop.dbus.exceptions.DBusExecutionException: Failed to find a compatible device for this connection
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
```

So it needs to be a little more complicated than what I wrote before but I think the principle should still stand...

Reply:
This seems to be working:

```java
   private void enableInterface(String deviceId, NetworkProperties properties, Optional<Device> device,
            NMDeviceType deviceType) throws DBusException {
        if (device.isPresent()) {
            enableInterface(deviceId, properties, device.get(), deviceType);
        } else {
            enableVirtualInterface(deviceId, properties, deviceType);
        }
    }

    private void enableVirtualInterface(String deviceId, NetworkProperties properties, NMDeviceType deviceType)
            throws DBusException {
        Map<String, Map<String, Variant<?>>> newConnectionSettings = NMSettingsConverter.buildSettings(properties,
                Optional.empty(), deviceId, deviceId, deviceType);

        Settings settings = this.dbusConnection.getRemoteObject(NM_BUS_NAME, NM_SETTINGS_BUS_PATH, Settings.class);
        DBusPath createdConnectionPath = settings.AddConnection(newConnectionSettings);
        Connection createdConnection = this.dbusConnection.getRemoteObject(NM_BUS_NAME, createdConnectionPath.getPath(),
                Connection.class);
        Device device = this.networkManager.getDeviceByIpIface(deviceId);

        try {
            this.networkManager.activateConnection(createdConnection, device);
        } catch (DBusExecutionException e) {
            logger.warn("Couldn't complete activation of {} interface, caused by:", deviceId, e);
        }
    }

    private void enableInterface(String deviceId, NetworkProperties properties, Device device, NMDeviceType deviceType)
            throws DBusException {
   // This is the old method without changes
```

Reply:
**Note**: This change also addresses the issues reported [by Sonar](https://github.com/eclipse/kura/pull/4816/checks?check_run_id=16418938420) so we're kinda killing two birds with one stone.